File: old/actions/install_sins_dns.py
"""
Install SiNS DNS server action
"""
import logging
import base64
import json
from cli import FileOps, SystemCtl
from cli.apt import Apt
from .base import Action
logger = logging.getLogger(__name__)

class InstallSinsDnsAction(Action):
    """Action to install SiNS DNS server"""
    description = "sins dns installation"

    def execute(self) -> bool:
        """Install SiNS DNS server from Debian package"""
        if not self.ssh_service or not self.apt_service:
            logger.error("SSH service or APT service not initialized")
            return False
        # Add Gemfury APT repository manually (matching Dockerfile format)
        logger.info("Adding Gemfury APT repository...")
        # Add repository source with [trusted=yes] flag (no GPG key needed)
        repo_source = "deb [trusted=yes] https://judyalvarez@apt.fury.io/judyalvarez /"
        add_source_cmd = f"echo '{repo_source}' | tee /etc/apt/sources.list.d/fury.list"
        output, exit_code = self.ssh_service.execute(add_source_cmd, sudo=True, timeout=30)
        if exit_code is not None and exit_code != 0:
            logger.error("Failed to add Gemfury repository source: %s", output[-200:] if output else "No output")
            return False
        # Update apt and install sins from repository
        logger.info("Installing SiNS DNS server from APT repository...")
        install_cmd = "apt-get update && apt-get install -y sins"
        output, exit_code = self.ssh_service.execute(install_cmd, sudo=True, timeout=300)
        if exit_code is not None and exit_code != 0:
            logger.error("Failed to install sins package: %s", output[-200:] if output else "No output")
            return False
        # Verify installation
        verify_cmd = "command -v sins >/dev/null && echo installed || echo not_installed"
        verify_output, verify_exit_code = self.ssh_service.execute(verify_cmd, sudo=True)
        if verify_exit_code != 0 or "installed" not in verify_output:
            logger.error("SiNS package was not installed correctly")
            return False
        # Get PostgreSQL connection info from config first, then container params, then defaults
        params = self.container_cfg.params if hasattr(self.container_cfg, "params") else {}
        # Use config postgres_host if available (from environment), otherwise use container params, then default
        postgres_host = None
        if self.cfg and hasattr(self.cfg, "postgres_host") and self.cfg.postgres_host:
            postgres_host = self.cfg.postgres_host
        elif params.get("postgres_host"):
            postgres_host = params.get("postgres_host")
        else:
            postgres_host = "10.11.3.18"  # Fallback default
        postgres_port = params.get("postgres_port", 5432)
        postgres_db = params.get("postgres_db", "dns_server")
        postgres_user = params.get("postgres_user", "postgres")
        postgres_password = params.get("postgres_password", "postgres")
        dns_port = params.get("dns_port", 53)
        web_port = params.get("web_port", 80)
        # Create PostgreSQL database if it doesn't exist
        logger.info("Ensuring PostgreSQL database '%s' exists...", postgres_db)
        # Install postgresql-client if not already installed (needed for psql)
        install_pg_client_cmd = "command -v psql >/dev/null || apt-get install -y postgresql-client"
        self.ssh_service.execute(install_pg_client_cmd, sudo=True, timeout=60)
        # Create database (ignore error if it already exists)
        create_db_cmd = f"PGPASSWORD={postgres_password} psql -h {postgres_host} -p {postgres_port} -U {postgres_user} -d postgres -tc \"SELECT 1 FROM pg_database WHERE datname = '{postgres_db}'\" | grep -q 1 || PGPASSWORD={postgres_password} psql -h {postgres_host} -p {postgres_port} -U {postgres_user} -d postgres -c \"CREATE DATABASE {postgres_db};\""
        output, exit_code = self.ssh_service.execute(create_db_cmd, sudo=True, timeout=30)
        if exit_code == 0:
            logger.info("PostgreSQL database '%s' is ready", postgres_db)
        else:
            logger.warning("Could not verify/create PostgreSQL database (may already exist): %s", output[-100:] if output else "No output")
        # Create appsettings.json
        logger.info("Configuring SiNS application settings...")
        # Generate a secure 256-bit (32 bytes) JWT secret key
        import secrets
        jwt_secret = secrets.token_urlsafe(32)  # 32 bytes = 256 bits
        appsettings = {
            "ConnectionStrings": {
                "DefaultConnection": f"Host={postgres_host};Port={postgres_port};Database={postgres_db};Username={postgres_user};Password={postgres_password}"
            },
            "DnsSettings": {
                "Port": dns_port
            },
            "WebSettings": {
                "Port": web_port
            },
            "Jwt": {
                "Key": jwt_secret,
                "Issuer": "SiNS-DNS-Server",
                "Audience": "SiNS-DNS-Client",
                "ExpirationMinutes": 1440
            }
        }
        appsettings_json = json.dumps(appsettings, indent=2)
        appsettings_b64 = base64.b64encode(appsettings_json.encode()).decode()
        # Determine appsettings location (check common locations for Debian package)
        # Try /etc/sins/ first, then /opt/sins/app/, then /usr/lib/sins/
        logger.info("Configuring SiNS appsettings.json...")
        # Check where the package installed the application
        find_app_cmd = "find /usr /opt /etc -name 'sins.dll' -o -name 'sins' -type f 2>/dev/null | head -1"
        app_location, _ = self.ssh_service.execute(find_app_cmd, sudo=True)
        # Write to all possible config locations to ensure correct config is used
        # .NET apps check /etc first, then working directory, so we write to all to overwrite any wrong configs
        config_locations = [
            "/etc/sins/appsettings.json",  # Primary location - .NET apps check /etc first
            "/opt/sins/appsettings.json",  # Service WorkingDirectory is /opt/sins
            "/opt/sins/app/appsettings.json",
        ]
        logger.info("Writing SiNS appsettings.json to all config locations...")
        config_written = False
        for config_path in config_locations:
            # Create directory if needed
            config_dir = "/".join(config_path.split("/")[:-1])
            mkdir_cmd = f"mkdir -p {config_dir}"
            self.ssh_service.execute(mkdir_cmd, sudo=True)
            # Write appsettings (overwrite any existing file)
            appsettings_cmd = f"echo {appsettings_b64} | base64 -d > {config_path}"
            output, exit_code = self.ssh_service.execute(appsettings_cmd, sudo=True)
            if exit_code == 0:
                logger.info("SiNS appsettings.json written to %s", config_path)
                config_written = True
            else:
                logger.warning("Failed to write appsettings.json to %s: %s", config_path, output[-100:] if output else "No output")
        if not config_written:
            logger.error("Failed to write appsettings.json to any location")
            return False
        return True

File: src/actions/install_sins_dns.go
package actions

import (
	"crypto/rand"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"strings"
	"enva/libs"
	"enva/services"
)

// InstallSinsDnsAction installs SiNS DNS server
type InstallSinsDnsAction struct {
	*BaseAction
}

func NewInstallSinsDnsAction(sshService *services.SSHService, aptService *services.APTService, pctService *services.PCTService, containerID *string, cfg *libs.LabConfig, containerCfg *libs.ContainerConfig) Action {
	return &InstallSinsDnsAction{
		BaseAction: &BaseAction{
			SSHService:   sshService,
			APTService:   aptService,
			PCTService:   pctService,
			ContainerID: containerID,
			Cfg:          cfg,
			ContainerCfg: containerCfg,
		},
	}
}

func (a *InstallSinsDnsAction) Description() string {
	return "sins dns installation"
}

func (a *InstallSinsDnsAction) Execute() bool {
	if a.SSHService == nil || a.APTService == nil {
		libs.GetLogger("install_sins_dns").Error("SSH service or APT service not initialized")
		return false
	}
	libs.GetLogger("install_sins_dns").Info("Adding Gemfury APT repository...")
	repoSource := "deb [trusted=yes] https://judyalvarez@apt.fury.io/judyalvarez /"
	addSourceCmd := fmt.Sprintf("echo '%s' | tee /etc/apt/sources.list.d/fury.list", repoSource)
	timeout := 30
	output, exitCode := a.SSHService.Execute(addSourceCmd, &timeout, true) // sudo=True
	if exitCode != nil && *exitCode != 0 {
		outputLen := len(output)
		start := 0
		if outputLen > 200 {
			start = outputLen - 200
		}
		libs.GetLogger("install_sins_dns").Error("Failed to add Gemfury repository source: %s", output[start:])
		return false
	}
	libs.GetLogger("install_sins_dns").Info("Installing SiNS DNS server from APT repository...")
	installCmd := "apt-get update && apt-get install -y sins"
	timeout = 300
	output, exitCode = a.SSHService.Execute(installCmd, &timeout, true) // sudo=True
	if exitCode != nil && *exitCode != 0 {
		outputLen := len(output)
		start := 0
		if outputLen > 200 {
			start = outputLen - 200
		}
		libs.GetLogger("install_sins_dns").Error("Failed to install sins package: %s", output[start:])
		return false
	}
	verifyCmd := "command -v sins  && echo installed || echo not_installed"
	verifyOutput, verifyExitCode := a.SSHService.Execute(verifyCmd, nil, true) // sudo=True
	if verifyExitCode == nil || *verifyExitCode != 0 || !strings.Contains(verifyOutput, "installed") {
		libs.GetLogger("install_sins_dns").Error("SiNS package was not installed correctly")
		return false
	}
	params := make(map[string]interface{})
	if a.ContainerCfg != nil && a.ContainerCfg.Params != nil {
		params = a.ContainerCfg.Params
	}
	postgresHost := "10.11.3.18"
	if a.Cfg != nil && a.Cfg.PostgresHost != nil {
		postgresHost = *a.Cfg.PostgresHost
	} else if p, ok := params["postgres_host"].(string); ok {
		postgresHost = p
	}
	postgresPort := 5432
	if p, ok := params["postgres_port"].(int); ok {
		postgresPort = p
	}
	postgresDB := "dns_server"
	if d, ok := params["postgres_db"].(string); ok {
		postgresDB = d
	}
	postgresUser := "postgres"
	if u, ok := params["postgres_user"].(string); ok {
		postgresUser = u
	}
	postgresPassword := "postgres"
	if p, ok := params["postgres_password"].(string); ok {
		postgresPassword = p
	}
	dnsPort := 53
	if p, ok := params["dns_port"].(int); ok {
		dnsPort = p
	}
	webPort := 80
	if p, ok := params["web_port"].(int); ok {
		webPort = p
	}
	libs.GetLogger("install_sins_dns").Info("Ensuring PostgreSQL database '%s' exists...", postgresDB)
	installPgClientCmd := "command -v psql  || apt-get install -y postgresql-client"
	timeout = 60
	a.SSHService.Execute(installPgClientCmd, &timeout, true) // sudo=True
	createDBCmd := fmt.Sprintf("PGPASSWORD=%s psql -h %s -p %d -U %s -d postgres -tc \"SELECT 1 FROM pg_database WHERE datname = '%s'\" | grep -q 1 || PGPASSWORD=%s psql -h %s -p %d -U %s -d postgres -c \"CREATE DATABASE %s;\"", postgresPassword, postgresHost, postgresPort, postgresUser, postgresDB, postgresPassword, postgresHost, postgresPort, postgresUser, postgresDB)
	timeout = 30
	output, exitCode = a.SSHService.Execute(createDBCmd, &timeout, true) // sudo=True
	if exitCode != nil && *exitCode == 0 {
		libs.GetLogger("install_sins_dns").Info("PostgreSQL database '%s' is ready", postgresDB)
	} else {
		outputLen := len(output)
		start := 0
		if outputLen > 100 {
			start = outputLen - 100
		}
		libs.GetLogger("install_sins_dns").Warning("Could not verify/create PostgreSQL database (may already exist): %s", output[start:])
	}
	libs.GetLogger("install_sins_dns").Info("Configuring SiNS application settings...")
	jwtSecret := generateJWTSecret()
	appsettings := map[string]interface{}{
		"ConnectionStrings": map[string]string{
			"DefaultConnection": fmt.Sprintf("Host=%s;Port=%d;Database=%s;Username=%s;Password=%s", postgresHost, postgresPort, postgresDB, postgresUser, postgresPassword),
		},
		"DnsSettings": map[string]int{
			"Port": dnsPort,
		},
		"WebSettings": map[string]int{
			"Port": webPort,
		},
		"Jwt": map[string]interface{}{
			"Key":              jwtSecret,
			"Issuer":           "SiNS-DNS-Server",
			"Audience":         "SiNS-DNS-Client",
			"ExpirationMinutes": 1440,
		},
	}
	appsettingsJSON, _ := json.MarshalIndent(appsettings, "", "  ")
	appsettingsB64 := base64.StdEncoding.EncodeToString(appsettingsJSON)
	configLocations := []string{
		"/etc/sins/appsettings.json",
		"/opt/sins/appsettings.json",
		"/opt/sins/app/appsettings.json",
	}
	libs.GetLogger("install_sins_dns").Info("Writing SiNS appsettings.json to all config locations...")
	configWritten := false
	for _, configPath := range configLocations {
		parts := strings.Split(configPath, "/")
		configDir := strings.Join(parts[:len(parts)-1], "/")
		mkdirCmd := fmt.Sprintf("mkdir -p %s", configDir)
		a.SSHService.Execute(mkdirCmd, nil, true) // sudo=True
		appsettingsCmd := fmt.Sprintf("echo %s | base64 -d > %s", appsettingsB64, configPath)
		output, exitCode = a.SSHService.Execute(appsettingsCmd, nil, true) // sudo=True
		if exitCode != nil && *exitCode == 0 {
			libs.GetLogger("install_sins_dns").Info("SiNS appsettings.json written to %s", configPath)
			configWritten = true
		} else {
			outputLen := len(output)
			start := 0
			if outputLen > 100 {
				start = outputLen - 100
			}
			libs.GetLogger("install_sins_dns").Warning("Failed to write appsettings.json to %s: %s", configPath, output[start:])
		}
	}
	if !configWritten {
		libs.GetLogger("install_sins_dns").Error("Failed to write appsettings.json to any location")
		return false
	}
	return true
}

func generateJWTSecret() string {
	// Generate a secure 256-bit (32 bytes) JWT secret key
	// In Go, use crypto/rand for secure random generation
	b := make([]byte, 32)
	rand.Read(b)
	return base64.URLEncoding.EncodeToString(b)
}

